### PR TITLE
Docs and templates

### PR DESCRIPTION
**Pull request by azdagron**
Signed-off-by: Andrew Harding <aharding@vmware.com>

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+| Host Service | Versions | Description |
+| ------------ | -------- | ----------- |
+| IdentityProvider | [v1](proto/spire/hostservice/server/identityprovider/v1/identityprovider.proto) | Provides an identity and bundle information |
+| AgentStore       | [v1](proto/spire/hostservice/server/agentstore/v1/agentstore.proto)             | Provides information about attested agents  |
+
+
+## Authoring Plugins
+
+For guidance in authoring a plugin see [AUTHORING](/docs/AUTHORING.md).
```
Comment (amartinezfayo) [inline README.md:69]:
```suggestion
For guidance in authoring a plugin, see [AUTHORING](/docs/AUTHORING.md).
```

File: README.md [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+## Migrating Pre-SDK Plugins
+
+To migrate existing pre-SDK plugins see [MIGRATING](/docs/MIGRATING.md).
```
Comment (amartinezfayo) [inline README.md:73]:
```suggestion
To migrate existing pre-SDK plugins, see [MIGRATING](/docs/MIGRATING.md).
```

File: README.md [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+## Versioning
+
+This repository is tagged along with SPIRE releases with the same name, even if
+there are no changes to the APIs between SPIRE versions. This allows consumers
+to always pick a tag that matches up with their deployment. Even so, SPIRE
+maintains API compatibility between SPIRE versions. SPIRE will clearly indicate
+in the CHANGELOG when APIs are deprecated and issue warnings at runtime when
```
Comment (amartinezfayo) [inline README.md:81]:
```suggestion
in the [CHANGELOG](https://github.com/spiffe/spire/blob/main/CHANGELOG.md) when APIs are deprecated and issue warnings at runtime when
```

File: docs/AUTHORING.md [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+    UpstreamAuthority "disk" {
+        plugin_data {
+            key_file_path = "some.key"
+            cert_file_path = "some.crt"
+        }
+    }
+```
+
+In order to receive this configuration data, as well as other core
+configurables, a plugin implements the [Config](proto/spire/service/common/config) service:
```
Comment (amartinezfayo) [inline docs/AUTHORING.md:26]: The link will not work, as it's relative to the docs dir.

File: docs/AUTHORING.md [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+import configv1 "github.com/spiffe/spire-plugin-sdk/proto/spire/service/common/config/v1"
+
+...
+
+type Plugin struct {
+    configv1.UnimplementedConfigServer
+}
+```
+
+- Implement the Configure RPC:
```
Comment (amartinezfayo) [inline docs/AUTHORING.md:47]:
```suggestion
- Implement the `Configure` RPC:
```

File: docs/AUTHORING.md [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+configurables, a plugin implements the [Config](proto/spire/service/common/config) service:
+
+Implementing this service is **optional**; plugins which do not require
+additional configuration are free to not implement it. However, if SPIRE
+receives configuration data for a plugin and that plugin does NOT implement
+a configuration service, SPIRE will fail to load the plugin.
+
+To implement this service
+
+- Embed the UnimplementedConfigServer struct:
```
Comment (amartinezfayo) [inline docs/AUTHORING.md:35]:
```suggestion
- Embed the `UnimplementedConfigServer` struct:
```

File: docs/AUTHORING.md [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+```
+func (p *Plugin) SetLogger(logger hclog.Logger) {
+}
+```
+
+The passed in logger can be stored by the plugin for later use.
+
+## Consuming Host Services
+
+Plugins obtain Host Services clients by implementing the BrokerHostServices
```
Comment (amartinezfayo) [inline docs/AUTHORING.md:115]:
```suggestion
Plugins obtain Host Services clients by implementing the `BrokerHostServices`
```

File: docs/AUTHORING.md [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+function on the NeedsHostServices interface. The function is passed a broker
```
Comment (amartinezfayo) [inline docs/AUTHORING.md:116]:
```suggestion
function on the `NeedsHostServices` interface. The function is passed a broker
```

File: docs/CONTRIBUTING.md [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+changes.  Care must be taken to not push the `replace` directive change up to
+SPIRE.
+
+Once those changes have been merged and you are ready to consume them from
+SPIRE, run `go get github.com/spiffe/spire-plugin-sdk@<commit hash>` in the SPIRE
+repository. This will update `go.mod` in SPIRE to use the latest pseudo version
+with that commit.
+
+When cutting a SPIRE release, this repository is tagged with the SPIRE
+release version. The release branch in SPIRE is be updated to depend explicitly
```
Comment (amartinezfayo) [inline docs/CONTRIBUTING.md:48]: SPIRE _needs_ to be updated?

File: docs/MIGRATING.md [hunk 10]
```diff
@@ -0,0 +1,149 @@
+# Migrating Plugins
+
+This document gives migration advice for porting existing plugins to the new
+SDK interfaces.
+
+## General Guidance
+
+### Configuration
+
+Pre-SDK plugin interfaces each contained their own Configure RPC. This RPC has
```
Comment (amartinezfayo) [inline docs/MIGRATING.md:10]:
```suggestion
Pre-SDK plugin interfaces each contained their own `Configure` RPC. This RPC has
```

File: docs/MIGRATING.md [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+been replaced with a specific, versioned
+[Config](/proto/spire/service/common/config) service.
+
+See the relevant section in [AUTHORING](AUTHORING.md) for implementation guidance.
+
+### GetPluginInfo
+
+Pre-SDK plugin interfaces each included an unused GetPluginInfo RPC. This
```
Comment (amartinezfayo) [inline docs/MIGRATING.md:18]:
```suggestion
Pre-SDK plugin interfaces each included an unused `GetPluginInfo` RPC. This
```

File: docs/MIGRATING.md [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+Migrated plugins _SHOULD_ return proper gRPC statuses when an error is
+encountered, e.g.:
+
+```
+return status.Error(codes.InvalidArgument, "blah")
+```
+
+### Consuming Host Services
+
+Plugin obtain Host Services clients by implementing the BrokerHostServices
```
Comment (amartinezfayo) [inline docs/MIGRATING.md:37]:
```suggestion
Plugins obtain Host Services clients by implementing the `BrokerHostServices`
```

File: docs/MIGRATING.md [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+function on the NeedsHostServices interface. The function is passed a broker
```
Comment (amartinezfayo) [inline docs/MIGRATING.md:38]:
```suggestion
function on the `NeedsHostServices` interface. The function is passed a broker
```

File: docs/MIGRATING.md [hunk 14]
```diff
@@ -0,0 +1,149 @@
...
+    if !broker.BrokerClient(&p.barClient) {
+        p.log.Warn("Bar host service is not implemented")
+    }
+    return nil
+}
+```
+
+As before, plugin authors can decide if the lack of support for a specific host
+service is an error or not. If the plugin returns an error from
+BrokerHostServices, the plugin will fail to load.
```
Comment (amartinezfayo) [inline docs/MIGRATING.md:80]:
```suggestion
`BrokerHostServices`, the plugin will fail to load.
```

File: docs/MIGRATING.md [hunk 15]
```diff
@@ -0,0 +1,149 @@
...
+
+The `Attest` RPC response returns selector values only. The selector type
+is inferred by SPIRE from the name of the plugin.
+
+### Server KeyManager
+
+RSA 1024 bit key support has been dropped.
+
+Keys managed by the KeyManager now have a required `fingerprint` component. The
+fingerprint is returned when keys are created or retrieved and by the
```
Comment (amartinezfayo) [inline docs/MIGRATING.md:117]:
```suggestion
fingerprint is returned when keys are created or retrieved by the
```

File: docs/MIGRATING.md [hunk 16]
```diff
@@ -0,0 +1,149 @@
...
+`SignData` RPC. It is used by SPIRE to detect if the key used in a signing
+operation has changed from underneath SPIRE.
+
+The PSSOptions have been moved inside of the SignDataRequest to couple them
```
Comment (amartinezfayo) [inline docs/MIGRATING.md:121]:
```suggestion
The `PSSOptions` struct have been moved inside of the `SignDataRequest` struct to couple them
```

**Review by amartinezfayo — COMMENTED** [review]
This is an awesome work @azdagron!
While I'm looking deeper and trying to exercise the guides, I thought to send a first round of quick suggestions and fixes.

File: README.md [hunk 17]
```diff
@@ -0,0 +1,91 @@
+This repository contains the service definitions, code generated stubs, and
```
[the remark this answers is not in the document]
Comment (azdagron) [reply README.md:1]: :+1:

File: README.md [hunk 18]
```diff
@@ -0,0 +1,91 @@
...
+| UpstreamAuthority | [v1](proto/spire/plugin/server/upstreamauthority/v1/upstreamauthority.proto) | Plugs SPIRE into an upstream PKI                      | [link](templates/server/upstreamauthority) |
+
+
+## Services
+
+### Common
+
+| Service | Versions | Description |
+| ------- | -------- | ----------- |
+| Config | [v1](proto/spire/service/common/config/v1/config.proto) | Used by SPIRE to configure the plugin. |
```
[the remark this answers is not in the document]
Comment (azdagron) [reply README.md:47]: I'll consistently add periods

File: README.md (hunk 1 above)
Comment (azdagron) [reply README.md:69]: :+1:

File: README.md (hunk 2 above)
Comment (azdagron) [reply README.md:73]: :+1:

File: README.md (hunk 3 above)
Comment (azdagron) [reply README.md:81]: :+1:

File: docs/AUTHORING.md (hunk 4 above)
Comment (azdagron) [reply docs/AUTHORING.md:26]: Fixed.

File: docs/AUTHORING.md (hunk 5 above)
Comment (azdagron) [reply docs/AUTHORING.md:47]: :+1:

File: docs/AUTHORING.md (hunk 6 above)
Comment (azdagron) [reply docs/AUTHORING.md:35]: :+1:

File: docs/AUTHORING.md (hunk 7 above)
Comment (azdagron) [reply docs/AUTHORING.md:115]: :+1:

File: docs/AUTHORING.md (hunk 8 above)
Comment (azdagron) [reply docs/AUTHORING.md:116]: :+1:

File: docs/CONTRIBUTING.md (hunk 9 above)
Comment (azdagron) [reply docs/CONTRIBUTING.md:48]: removed _be_

File: docs/MIGRATING.md (hunk 10 above)
Comment (azdagron) [reply docs/MIGRATING.md:10]: :+1:

File: docs/MIGRATING.md (hunk 11 above)
Comment (azdagron) [reply docs/MIGRATING.md:18]: :+1:

File: docs/MIGRATING.md (hunk 12 above)
Comment (azdagron) [reply docs/MIGRATING.md:37]: :+1:

File: docs/MIGRATING.md (hunk 14 above)
Comment (azdagron) [reply docs/MIGRATING.md:80]: :+1:

File: docs/MIGRATING.md (hunk 13 above)
Comment (azdagron) [reply docs/MIGRATING.md:38]: :+1:

File: docs/MIGRATING.md (hunk 15 above)
Comment (azdagron) [reply docs/MIGRATING.md:117]: :+1:

File: docs/MIGRATING.md (hunk 16 above)
Comment (azdagron) [reply docs/MIGRATING.md:121]: :+1:

**Review by amartinezfayo — APPROVED** [review]
Thanks for doing this @azdagron! I didn't find anything else to change at this time.
Awesome work!

**Comment by azdagron** [comment]
:o i lucked out!

**Pull request merged**